torch/distributed/algorithms/ddp_comm_hooks/mixed_precision_hooks.py

```diff
@@ -0,0 +1,82 @@
+
+from dataclasses import dataclass
+from typing import Any
+
+@dataclass
+class _AllreduceUpcastHookState:
+    """
+    State to manage DDP mixed precision in backward / gradient communication.
+    This contains a weakref to the DDP module for access to reducer and process
+    group, and a stream to run parameter and gradient upcasts.
+    """
+    ddp_weakref: Any
+    upcast_stream: torch.cuda.Stream
+    upcast_params_and_grads: bool
+    wait_for_stream_enqueued: bool = False
+
+def _reducer_allreduce_and_upcast_hook(
+    hook_state: _AllreduceUpcastHookState, bucket: dist.GradBucket
+) -> torch.futures.Future[torch.Tensor]:
+    """
+    Performs allreduce in the reduced precision given by DDP's mixed precision
+    reduce_dtype, and upcasts parameters and gradients to fp32 in preparation
+    to run the optimizer.
+    """
+    ddp_weakref = hook_state.ddp_weakref
+    reducer, process_group = ddp_weakref().reducer, ddp_weakref().process_group
+    gradient_is_bucket_view = ddp_weakref().gradient_as_bucket_view
+    # Cast bucket if different than param_dtype.
+    if (
+        ddp_weakref().mixed_precision.param_dtype != ddp_weakref().mixed_precision.reduce_dtype
+    ):
+        # Cast bucket tensor to reduce_dtype
+        bucket.set_buffer(bucket.buffer().to(ddp_weakref().mixed_precision.reduce_dtype))
+    fut = reducer._run_allreduce_hook(bucket)
+    ret_fut = torch.futures.Future()
+    stream = hook_state.upcast_stream
+    with torch.cuda.stream(stream):
+        fut.wait()
+        bucket.buffer().div_(process_group.size())
+        ret_fut.set_result(bucket.buffer())
+
+        if hook_state.upcast_params_and_grads:
+            # Upcast parameters and gradients so optimizer step can run in fp32.
+            params, grads = bucket.parameters(), bucket.gradients()
+            for p, g in zip(params, grads):
+                p.data = p._fp_param
+                if not gradient_is_bucket_view:
+                    g.data = g.data.to(p.data.dtype)
+                    p.grad = g
+                else:
+                    p.grad.data = p.grad.to(p.data.dtype)
+
+    # enqueue a callback to wait for this stream at end of backward
+    def wait_for_stream_cb():
+        torch.cuda.current_stream().wait_stream(stream)
+        # Remove post-backward hooks since they are re-installed in next
+        # iteration, similar to FSDP.
+        # Parameters that don't require grad still needed to be casted since
+        # they may participate in computation. However, they would not be recast
+        # by hook above as they don't have a grad hook installed, so cast them
+        # back here.
+        for n, p in ddp_weakref().module.named_parameters():
+            if hasattr(p, '_ddp_mp_hook_state'):
+                p._ddp_mp_hook_state[1].remove()
+                delattr(p, '_ddp_mp_hook_state')
+            if not p.requires_grad and not hasattr(p, '_ddp_ignored'):
+                p.data = p._fp_param
```

Comment:
the upcast has already been done in above codes?

Reply:
This is for params that don't require grad. They don't get the above code run since their autograd hooks do not fire, but we still need to upcast them as they were downcast before the forward (we need to do that since they may participate in computation)